### ISSUE TITLE
Add functionlity CPU micrcode updates

### ISSUE THREAD

**Issue by rn** [linuxkit/linuxkit#2843]
For baremetal deployments we should have a package which updates the CPU microcode either from the debian or the alpine base packages. It likely only requires dumping the microcode binary in the right place in the rootfs and the kernel will take care of the rest.

**Comment by rn** [comment]
A bit more prodding around, there are three ways to upgrade the CPU microcode (see [here](https://www.kernel.org/doc/Documentation/x86/early-microcode.txt))
- via `/dev/cpu/microcode`
- via `/sys/devices/system/cpu/microcode/reload`
- via prepended initrd

The first two may happen too late in the boot process. In fact, I build an image with the FW files in `/lib/firmware/intel-ucode` and it did not get loaded on a packet.net type 1e machine, which was on revision `0x6a` with the lates firmware from Intel being `0xc2`.

The [`iucode-tool`](https://gitlab.com/iucode-tool/iucode-tool) has a mode to list details about ucode files (`iucode-tool -L .` in the `intel-ucode` directory downloaded from Intel). The `iucode-tool` also has an option to build a cpio archive with the latest microcode:

```
iucode_tool --normal-earlyfw --write-earlyfw=ucode.cpio ./intel-ucode
```

This `cpio` archive can be prepended to the initrd:
```
mv packet-initrd.img packet-initrd.img.orig
cat ucode.cpio packet-initrd.img.orig > packet-initrd.img
```

And when booted the first line of `dmesg` reads:
```
[    0.000000] microcode: microcode updated early to revision 0xc2, date = 2017-11-16
```

The AMD update works similarily

Bottom line is that the original suggestion of creating a `ucode` package and including it in the `YAML` file will not work.

My current thinking is to build the `cpio` archive as part of the kernel build and stash it in the kernel package (The intel `ucode.cpio` is 1.6MB, so negligible). The we add a `-baremetal` flag to `moby build`/`linuxkit build` which, when specified, will extract prepend the `cpio` archive in front of the `initrd` as shown above. Initially we only support it for building `kernel+initrd` but we should investigate how to add it to the other builds as well.

@justincormack @ijc any better ideas?

**Comment by ijc** [comment]
Yeah, it sucks but the "prepend to initrd" way is the only safe/reliable way to do things.

The `-baremetal` UX is a bit nasty, but it's about as good as it'll get (doing it later e.g. with a `linuxkit push baremetal` fails due to not having the yml with the hash to use at that point).

A fair few of the other formats are actually an extra pipeline stage after `kernel+initrd`, aren't they, so maybe we get a bunch of those "for free" (ish).

Anyway ack to your plan.

**Comment by rn** [comment]
The alternative is to always prepend if `ucode.cpio` is present in the kernel package. It's a small size increase (and I can probably make it smaller still as I used the `--normal-earlyfw` option and there is another option to make a smaller `cpio` archive). I'd need to experiment if this impacts booting on hypervisors (which it shouldn't).

**Comment by justincormack** [comment]
The way that the build currently works is that the basic output is a tarball with "special" files in `/boot`, currently `/boot/kernel` and `/boot/cmdline`, which the `initrd` builder splits out and omits from the initrd. These are put in there by the `kernel` section of the yaml. Other formats do different things with these, eg leave the kernel there and point the bootloader at it, on the ISO.

So adding a `ucode` section to the kernel config, putting that in `/boot/ucode` and then getting the build code to prepend that to the initrd (padded correctly) would seem correct.

**Comment by justincormack** [comment]
Alternatively maybe you can build it into the kernel image? There is an option to specify an initramfs at build time, which is normally empty. Then our one would still be appended I think.

**Comment by ijc** [comment]
> There is an option to specify an initramfs at build time, which is normally empty. Then our one would still be appended I think.

I think the bootloader provided initramfs is ignored when that option is is set (worth trying it, but I wouldn't wet on it working as we need here)

**Comment by ijc** [comment]
> I think the bootloader provided initramfs is ignored when that option is is set (worth trying it, but I wouldn't wet on it working as we need here)

I take it back -- it looks like the external initramfs is indeed unpacked over the internal one.

However it unfortunately looks like the early microcode stuff only considers the external initramfs AFAICT from the code..

**Comment by rn** [comment]
several bootloaders, such as `grub`, `syslinux` etc also allow specifying more than one initrd, so in grub, something like:
```
initrd /boot/microcode.cpio /boot/initrd.img
```
should work. But I'd prefer if we had solution for when we use an initrd so the concatenated initrds seem better

**Comment by rn** [comment]
via `CONFIG_EXTRA_FIRMWARE` we could also compile the firmware into the kernel...need to play with that a bit more. It would mean that all x86_64 kernels would have to pay the `ucode` tax of 1-2MB. I also don't know what the format needs to be. 

It's a little tedious, but plausible.

**Comment by sprat** [comment]
I just discovered that the ucode is only included when the output format is kernel+initrd, so I'll open a bug to address the other cases.
